File: 0051-n-queens/0051-n-queens.py
from typing import List

class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        board = [[False for _ in range(n)] for _ in range(n)]
        return self.queens(board, 0)

    def queens(self, board: List[List[bool]], row: int) -> List[List[str]]:
        if row == len(board):
            return self.boardConfig(board)
        
        ans = []
        
        for col in range(len(board[0])):
            if self.isSafe(board, row, col):
                board[row][col] = True  # Place queen
                ans.extend(self.queens(board, row + 1))  # Recursively solve for the next row
                board[row][col] = False  # Backtrack
        
        return ans

    def isSafe(self, board: List[List[bool]], row: int, col: int) -> bool:
        # Check column
        for i in range(row):
            if board[i][col]:
                return False

        # Check left diagonal
        i, j = row, col
        while i >= 0 and j >= 0:
            if board[i][j]:
                return False
            i -= 1
            j -= 1

        # Check right diagonal
        i, j = row, col
        while i >= 0 and j < len(board[0]):
            if board[i][j]:
                return False
            i -= 1
            j += 1

        return True

    def boardConfig(self, board: List[List[bool]]) -> List[List[str]]:
        result = []
        config = []
        
        for row in board:
            row_str = ''.join("Q" if cell else "." for cell in row)
            config.append(row_str)
        
        result.append(config)
        return result

File: 0051-n-queens/0051-n-queens.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        boolean[][] board = new boolean[n][n];

        return queens(board, 0);
    }

    List<List<String>> queens (boolean[][] board, int row){
        if (row == board.length) {
            return boardConfig(board);
        }

        List<List<String>> ans = new ArrayList<>();

        for (int col = 0; col < board[0].length; col++) {
            if (isSafe(board, row, col)) {
                board[row][col] = true;
                ans.addAll(queens(board, row+1));  // if it is safe to enter till last row then we are returning the board config here
                board[row][col] = false;     // reverting changes made to the board for other configs
            }
        }

        return ans;
    }

    boolean isSafe(boolean[][] board, int row, int col) {

        // checking for column
        for (int i = row; i >= 0; i--) {
            if (board[i][col]) {
                return false;
            }
        }

        // checking left diagonali
        int i = row;
        int j = col;
        while (i >= 0 && j >= 0) {
            if (board[i][j]) {
                return false;
            }
            i--;
            j--;
        }

        // checking for right diagonal
        i = row;
        j = col;
        while (i >= 0 && j <= board[0].length-1) {
            if (board[i][j]) {
                return false;
            }
            i--;
            j++;
        }

        return true;
    }

    List<List<String>> boardConfig (boolean[][] board) {
        List<List<String>> list = new ArrayList<>();
        List<String> config = new ArrayList<>();

        for (boolean[] arr : board) {
            StringBuilder sb = new StringBuilder();
            for (boolean ele : arr) {
                if (ele) {
                    sb.append("Q");
                }
                else {
                    sb.append(".");
                }
            }
            config.add(sb.toString());
        }

        list.add(config);
        return list;
    }
}